Italian language selection

zenseii opened the pull request:
This PR makes it possible to choose the Italian language from the Language selection options. Accents etc will be added in a separate PR.

ihhub commented:
Hi @zenseii , everything looks fine but I would insist to add a real code to at least generate by copy-pasting some letters like it was done for French language at first. In this case users will have all letters even if some of them are wrong.

ihhub commented on files/lang/it.po:
```diff
@@ -8,7 +8,7 @@ msgstr ""
 "Report-Msgid-Bugs-To: \n"
 "POT-Creation-Date: 2022-02-17 08:59+0100\n"
-"PO-Revision-Date: 2021-11-23 23:57+0100\n"
+"PO-Revision-Date: 2022-03-16 19:22+0100\n"
```
Could we please revert the changes in this file as they do absolutely nothing? :)

ihhub commented on src/fheroes2/agg/agg_image.cpp:
```diff
@@ -1401,6 +1401,230 @@ namespace
+
+        // Normal font.
+        {
+            std::vector<fheroes2::Sprite> & font = _icnVsSprite[ICN::FONT];
```
Isn't this code exactly the same as for French alphabet? If yes why do we even generate it?

ihhub reviewed (commented):
Hi @zenseii , I left just 2 comments here for further discussion. Could you please take a look?

zenseii commented:
Hi @ihhub ! I've now made the first batch of small and normal fonts. All letters with diacritics are included and have been tested. I should still make a change and try to create a diacritic like grave accent `` ` `` once instead of for every special character in the small fonts. Many of the same diacritics are found in French, although some where missing and I have added them in Italian like the `È` and  `É`.

ihhub commented on src/fheroes2/agg/agg_image.cpp:
```diff
@@ -1401,6 +1401,251 @@ namespace
+        {
+            std::vector<fheroes2::Sprite> & font = _icnVsSprite[ICN::FONT];
+
+            font[6].resize( font[85].width(), font[85].height() + 3 );
```
This letter is not generated for small font.

zenseii replied:
This letter is generated to create the grave accent `` ` ``. This accent was a pure copy paste from generateFrenchAlphabet(). This is why font[6] is copied for all the characters with grave accent `` ` ``.

zenseii replied:
But yes, I should  just generate the grave accent once for small font and then use that for every special character, just like done for normal font by setting it to font[6].

ihhub replied:
Then why do we need font[6] in first place if we could generate it for letter 192-32 straight away?

zenseii replied:
We could wouldn't that mean that we would have to copy all this code below for every character that uses the grave accent `` ` ``

```cpp
            fheroes2::Copy( font[85], 0, 0, font[6], 0, 3, font[85].width(), font[85].height() );
            fheroes2::Copy( font[6], 2, 3, font[6], 4, 0, 1, 1 );
            fheroes2::Copy( font[6], 2, 4, font[6], 5, 0, 1, 1 );
            fheroes2::Copy( font[6], 2, 3, font[6], 5, 1, 1, 1 );
            fheroes2::Copy( font[6], 2, 4, font[6], 6, 1, 1, 1 );
            fheroes2::Copy( font[6], 3, 3, font[6], 7, 1, 1, 1 );   
```

Instead of just repeating this line:
```cpp
            fheroes2::Copy( font[6], 4, 0, font[192 - 32], 6, 0, 4, 2 );
```

ihhub replied:
No, We can apply the change for 192-32 and then copy from 192-32 for other characters.

zenseii replied:
Aha I see. So generate it for `À` in this case, and then copy only the `` ` `` part of that `font[192-32]`.

ihhub replied:
That's right. We will avoid having an extra image.

zenseii replied:
I'll try to have it implemented by this evening. One question I have is if the x and y coordinates for copying from a sprite are the usual where 0,0 is in the upper left corner?

ihhub replied:
Yes, it is a top-left corner.

ihhub commented on src/fheroes2/agg/agg_image.cpp:
```diff
@@ -1401,6 +1401,251 @@ namespace
+            fheroes2::Copy( font[6], 4, 0, font[217 - 32], 5, 0, 4, 2 );
+            fheroes2::updateShadow( font[217 - 32], { -1, 2 }, 2 );
+
+            // a with grave accent `.
```
Based on provided by you screenshots Italian font contains small letters at positions 96 - 100 (Discord server, translations channel, 18th March 2022). Why are they different here?

zenseii replied:
Hi @ihhub ! Basically I found out that the game does indeed use CP1252 in the code, even if the resources as seen in the screenshot you mentioned are from 96-100. Here is that screenshot:
https://cdn.discordapp.com/attachments/903468433402900480/954119761598513152/unknown.png

However, I could not find any place where apart from UI where these where used, so in no Italian original map files. I managed to show these letters in the original game by using a hexeditor and inserting these special characters, using their CP1252 values. I then did another test by editing the game in ram using cheat engine and found that I could show the special letters using CP1252 values for them.

In other words, I'm using these values in order to support the text in the Italian map files. I haven't actually found any instance of them in any maps, but I haven't looked through all the text in all map and campaign files.

The logical choice for me was to use CP1252 values to maintain support for the original game in case any maps use the special characters.

Hex-edited CP1252 values in map description:
![image](https://user-images.githubusercontent.com/12501091/160010817-663525dd-aefe-46d7-ad61-42675ff28c54.png)


Original game showing this description (note the original game's resources does not contain the special characters `ÀÈÉÌÒÙé` which is why they are all replaced with `ù`):
![image](https://user-images.githubusercontent.com/12501091/160010521-104d6e3c-d8a3-47cf-a12b-5e5df4f2938e.png)

ihhub replied:
Got it. Thanks for a detailed explanation!

ihhub reviewed (changes requested):
Hi @zenseii , I left few comments here. Also with these changes do you see that the description of maps is still valid? As we found out Italian font doesn't have letters related to CP1252.

zenseii commented:
@ihhub I have addressed your comments and this PR is ready for review when you have the opportunity. I would like to ask that you look at the generation of the small font lower case `ì` and check if it is done in an acceptable way. 

I also fixed the issue of the it.po having a `\v` as one user reported giving a warning during compilation.

ihhub commented on src/fheroes2/agg/agg_image.cpp:
```diff
@@ -1401,6 +1401,229 @@ namespace
+            font[236 - 32] = font[73];
+            fheroes2::FillTransform( font[236 - 32], 0, 0, font[236 - 32].width(), 2, 1 );
+            fheroes2::Copy( font[192 - 32], 7, 0, font[236 - 32], 1, 0, 4, 2 );
+            fheroes2::updateShadow( font[236 - 32], { -1, 1 }, 2 );
```
I think this is a mistake: shadow offset must be the same for every letter.

zenseii replied:
Yes, you're right, it's a human error 😅

ihhub reviewed (changes requested):
Hi @zenseii , I left one more comment. Could you please take a look?

ihhub commented:
@zenseii , huge thanks for your changes!
